server/src/main/java/org/opensearch/search/backpressure/SearchBackpressureManager.java

```diff
@@ -0,0 +1,250 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.search.backpressure;
+
+import com.sun.management.OperatingSystemMXBean;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.action.search.SearchShardTask;
+import org.opensearch.common.Streak;
+import org.opensearch.common.SuppressForbidden;
+import org.opensearch.common.util.TokenBucket;
+import org.opensearch.monitor.jvm.JvmStats;
+import org.opensearch.search.backpressure.trackers.ResourceUsageTracker;
+import org.opensearch.tasks.CancellableTask;
+import org.opensearch.tasks.Task;
+import org.opensearch.tasks.TaskResourceTrackingService;
+import org.opensearch.tasks.TaskResourceTrackingService.TaskCompletionListener;
+import org.opensearch.threadpool.ThreadPool;
+
+import java.lang.management.ManagementFactory;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.DoubleSupplier;
+import java.util.function.LongSupplier;
+import java.util.stream.Collectors;
+
+/**
+ * SearchBackpressureManager is responsible for monitoring and cancelling in-flight search tasks if they are
+ * breaching resource usage limits when the node is in duress.
+ */
+@SuppressForbidden(reason = "OperatingSystemMXBean#getProcessCpuLoad")
+public class SearchBackpressureManager implements Runnable, TaskCompletionListener {
```

Comment:
This class is only performing cancellation on resource usage breaches. Should we call it `SearchCancellationService` instead?
This is more of a service than manager since it needs to run in the background?

Reply:
Makes sense; will update it.